samcli/lib/bootstrap/companion_stack/data_types.py

```diff
@@ -0,0 +1,127 @@
+"""
+    Date type classes for companion stacks
+"""
+import re
+from typing import Optional
+from samcli.lib.utils.hash import str_checksum
+
+
+class CompanionStack:
+    """
+    Abstraction class for the companion stack
+    Companion stack name will be generated by this class.
+    """
+
+    _parent_stack_name: str
+    _escaped_parent_stack_name: str
+    _parent_stack_hash: str
+    _stack_name: str
+
+    def __init__(self, parent_stack_name: str) -> None:
+        self._parent_stack_name = parent_stack_name
+        self._escaped_parent_stack_name = re.sub(r"[^a-z0-9]", "", self._parent_stack_name.lower())
+        self._parent_stack_hash = str_checksum(self._parent_stack_name)
+        self._stack_name = f"{self._parent_stack_name[:104]}-{self._parent_stack_hash[:8]}-CompanionStack"
```

Comment:
Why escaped? Why `104`? `8`?

Reply:
Escaped stack name is used for naming ECR repos.
There is max 128 characters limit on the length of stack name.
```
repo_name = 
    escaped_stack_name + stack_md5[:8] + "/" + escaped_lambda_logical_id + function_md5[:8] + "repo"
    #128        +         8      +  1  +     64      +        8      +      4 = 213 max char
```
104 for parent stack name, 8 for hashes in case of duplication, 16 for two dashes and "CompanionStack" suffix.